### PR TITLE
Redocument non-blocking Fibers and scheduler

### PR DESCRIPTION
**Pull request by zverok**
I propose new documentation approach for new features in Ruby 3.0: non-blocking Fiber and the scheduler.

Right now, the documentation is in a confusing state. First, the `doc/scheduler.md` is even not rendered correclty: https://docs.ruby-lang.org/en/master/doc/scheduler_md.html (relatively easy to fix: wrong markdown markup for the code-blocks), and Fiber class itself has no mention for new concepts, and no docs for new methods: https://docs.ruby-lang.org/en/master/Fiber.html#method-c-schedule.

But on the bigger level, the documentation is quite hard to follow unless you are already fully in context of asynchronous loops and schedulers.

I am trying to fix it by:

* adding to the `Fiber` class necessary docs, both high-level overview and particular method details
* redocumenting the expected scheduler interface via "imaginary" `Fiber::SchedulerInterface` class: it is present only in docs to leverage RDoc's method-by-method documentation, be able to link to them separately and so on

Test rendering of the docs on my personal site:

* [Fiber](https://zverok.github.io/ruby-rdoc/Fiber.html)
* [Fiber::SchedulerInterface](https://zverok.github.io/ruby-rdoc/Fiber/SchedulerInterface.html)

I have not yet in this PR removed doc/scheduler.md, but I suggest to, as it is completely superseded by new docs.

I'd be really grateful if @ioquatix will find some time to review this initiative in the upcoming days.

### CODE REVIEW

File: cont.c [hunk 1]
```diff
@@ -2196,12 +2306,44 @@ rb_fiber_transfer(VALUE fiber_value, int argc, const VALUE *argv)
...
+ *  call-seq:
+ *     Fiber.blocking? -> false or number
+ *
+ *  Returns +false+ if the current fiber is non-blocking.
+ *  Fiber is non-blocking if it was created via passing <tt>blocking: false</tt>
+ *  to Fiber.new, or via Fiber.schedule.
+ *
+ *  If the current Fiber is blocking, the method, unlike usual
+ *  predicate methods, returns a *number* of blocking fibers currently
+ *  running (TBD: always 1?).
```
Comment (zverok) [inline cont.c:2339]: @ioquatix Can you please elaborate? I can't guess neither by tests nor by code.

**Comment by ioquatix** [comment]
Sure, I can take a look. Can you confirm what parts of the markdown syntax are incorrect?

**Comment by zverok** [comment]
@ioquatix Thanks!

`~~~ruby` is not supported by RDoc Markdown renderer apparently, it should be three-backticks-ruby (compare with [doc/ractor.md](https://github.com/ruby/ruby/edit/master/doc/ractor.md) and how it is [rendered](https://docs.ruby-lang.org/en/master/doc/ractor_md.html) )

**Comment by ioquatix** [comment]
Oh, so RDoc Markdown is not common mark compatible?

Okay, I'll fix it.

**Comment by zverok** [comment]
@ioquatix Ping!

**Comment by ioquatix** [comment]
https://github.com/ruby/ruby/pull/3965 can you please check it.

**Comment by zverok** [comment]
@ioquatix How is `doc/fiber.md` is supposed to be related to `Fiber`'s class docs? Now it looks like a class docs replacement. (The previous version at least was more like "separate explanation for Scheduler's design and usage").
What's the supposed fate of the current (authored by me) PR?

**Comment by zverok** [comment]
cc @ko1 @marcandre

**Comment by marcandre** [comment]
If the issue is documentation duplication, that's a nice problem to have 😅 .  I'll look at this tomorrow if it's not resolved by then; too tired to think straight.

**Comment by marcandre** [comment]
@zverok I think it's great to have detailed API documentation and an overview documentation.
I propose to commit this as is; I will rephrase a bit and we can improve later (including the unanswered question).

**Comment by zverok** [comment]
Thanks @marcandre! And sorry for pushing it so hard.

**Comment by marcandre** [comment]
Very good work @zverok 
You did well to push 👍

**Comment by ioquatix** [comment]
Thanks everyone for your awesome effort writing documentation.
